### PR TITLE
Introduce Story NFT

### PR DESCRIPTION
**Pull request by sebsadface**
## Description

This PR introduces the Story NFT contracts, enabling developers to quickly bootstrap an NFT contract with built-in functionality for IP and derivative registration.

### Overview of New Interfaces and Contracts

- **`OrgNFT` & `IOrgNFT`**: Represents organizations deploying Story NFTs. Each organization is assigned a unique `OrgNFT` token, which establish a one-to-one mapping between the organization and its instance of a Story NFT contract.

- **`IStoryNFTFactory` & `StoryNFTFactory`**: Serve as the entry point for registering organizations and deploying new Story NFT contracts. The factory allows easy Story NFT creation and ensures proper linkage between organizations and their NFTs.

- **`IStoryNFT` & `BaseStoryNFT`**: An interface and abstract contract defining standard functions and behaviors for all Story NFTs. Custom Story NFTs needs to inherit these to ensure compatibility across different Story NFT contracts.

- **`IERC5192`**: A minimal interface for soulbound NFTs (non-transferable tokens), implemented by the `StoryBadgeNFT` contract.

- **`IStoryBadgeNFT` & `StoryBadgeNFT`**: Provide a default implementation of `IStoryNFT` and `BaseStoryNFT`, serving as the default template for Story NFTs created via the `StoryNFTFactory`.

### IP Relationships

Upon minting, each token in `OrgNFT` and instances of `StoryNFTs` is registered as an IP, with relationships defined as follows:

- The ancestor of all Story NFT IPs is the `OrgNFT` token of Story (token ID 0).

- All other `OrgNFT` tokens, representing different organizations, are considered derivatives of Story's Org token.

- Tokens minted through each `StoryNFT` contract instance are considered derivatives of the `OrgNFT` associated with that specific contract.

### Additional Changes

- Added deployment scripts for the newly introduced contracts.

## Test Plan

- Added test cases for each new function (located in `test/story-nft`).
- All existing and new tests pass locally.

### CODE REVIEW

File: contracts/story-nft/StoryNFTFactory.sol [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+            piLicenseTemplate: PILICENSE_TEMPLATE,
+            defaultLicenseTermsId: DEFAULT_LICENSE_TERMS_ID
+        });
+        rootStoryNft.initialize({
+            owner_: rootOrgOwner,
+            orgNft_: address(ORG_NFT),
+            orgTokenId_: 0,
+            orgIpId_: ORG_NFT.getRootOrgIpId(),
+            initData: rootStoryNftInitData
+        });
```
Comment (kingster-will) [inline contracts/story-nft/StoryNFTFactory.sol:121]: We might not need to deploy StoryBadgeNFT for Story Org when initialize the factory.
I feel we should deploy an empty StoryBadegeNFT as default template separately and only need to pass in the address of default template in the contractor function.
Then the initilaize function only need accessManager and signer.
The BadgeNFT for Story can be deployed later with deployByAdmin.

**Review by kingster-will — COMMENTED** [review]
Good PR!

File: contracts/interfaces/story-nft/IStoryBadge.sol [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.26;
+
+import { IERC721Metadata } from "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";
+
+import { IERC5192 } from "./IERC5192.sol";
+
+/// @title Story Badge Interface
+/// @notice A Story Badge is a soulbound NFT that has an unified token URI for all tokens.
+interface IStoryBadge is IERC5192, IERC721Metadata {
```
[the remark this answers is not in the document]
Comment (sebsadface) [reply contracts/interfaces/story-nft/IStoryBadge.sol]: `StoryBadgeNFT` already inherits `IERC4906` from `ERC721URIStorage`

File: contracts/interfaces/story-nft/IStoryNFTFactory.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    /// @dev Enforced to be only callable by the protocol admin in governance.
+    /// @param storyNftTemplate The address of a whitelisted StoryNFT template to be cloned.
+    /// @param storyNftOwner The address of the owner of the new StoryNFT contract.
+    /// @param orgName The name of the organization.
+    /// @param orgTokenURI The token URI of the organization NFT.
+    /// @param initData The initialization data for the StoryNFT (see {IStoryBadge-InitParams} for an example).
+    /// @return orgTokenId The token ID of the organization NFT.
+    /// @return orgIpId The ID of the organization IP.
+    /// @return storyNft The address of the dployed StoryNFT
+    function deployByAdmin(
```
[the remark this answers is not in the document]
Comment (sebsadface) [reply contracts/interfaces/story-nft/IStoryNFTFactory.sol]: Resolved! See: https://github.com/storyprotocol/protocol-periphery-v1/compare/725e774409b9137dabcc0776ea4e638640a03db8..61049137d2adb38b0949fbe7c258199d54d6df2e

File: contracts/story-nft/OrgNFT.sol [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    /// @param recipient The address of the recipient of the root organization token.
+    /// @param tokenURI_ The URI of the root organization token.
+    /// @return rootOrgTokenId The ID of the root organization token.
+    /// @return rootOrgIpId The ID of the root organization IP.
+    function mintRootOrgNft(
+        address recipient,
+        string memory tokenURI_
+    ) external onlyStoryNFTFactory returns (uint256 rootOrgTokenId, address rootOrgIpId) {
+        (rootOrgTokenId, rootOrgIpId) = _mintAndRegisterIp(recipient, tokenURI_);
+        _getOrgNFTStorage().rootOrgIpId = rootOrgIpId;
```
[the remark this answers is not in the document]
Comment (sebsadface) [reply contracts/story-nft/OrgNFT.sol]: Resolved! see: https://github.com/storyprotocol/protocol-periphery-v1/compare/725e774409b9137dabcc0776ea4e638640a03db8..61049137d2adb38b0949fbe7c258199d54d6df2e

File: contracts/story-nft/StoryBadge.sol [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        _signer = signer_;
+        emit StoryBadgeSingerUpdated(signer_);
+    }
+
+    /// @notice Updates the unified token URI for all badges.
+    /// @param tokenURI_ The new token URI.
+    function setTokenURI(string memory tokenURI_) external onlyOwner {
+        _tokenURI = tokenURI_;
+        emit StoryBadgeTokenURIUpdated(tokenURI_);
+    }
```
[the remark this answers is not in the document]
Comment (sebsadface) [reply contracts/story-nft/StoryBadge.sol]: Resolved! see: https://github.com/storyprotocol/protocol-periphery-v1/compare/725e774409b9137dabcc0776ea4e638640a03db8..61049137d2adb38b0949fbe7c258199d54d6df2e

File: contracts/story-nft/StoryBadge.sol [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+import { IERC721Metadata } from "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";
+import { MessageHashUtils } from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import { SignatureChecker } from "@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol";
+
+import { BaseStoryNFT } from "./BaseStoryNFT.sol";
+import { IStoryBadge } from "../interfaces/story-nft/IStoryBadge.sol";
+
+/// @title Story Badge
+/// @notice A Story Badge is a soulbound NFT that has an unified token URI for all tokens.
+contract StoryBadge is IStoryBadge, BaseStoryNFT, ERC721Holder {
```
[the remark this answers is not in the document]
Comment (sebsadface) [reply contracts/story-nft/StoryBadge.sol]: Resolved! see: https://github.com/storyprotocol/protocol-periphery-v1/compare/725e774409b9137dabcc0776ea4e638640a03db8..61049137d2adb38b0949fbe7c258199d54d6df2e

File: contracts/story-nft/StoryNFTFactory.sol (hunk 1 above)
Comment (sebsadface) [reply contracts/story-nft/StoryNFTFactory.sol]: Resolved, and added deploy script! see: https://github.com/storyprotocol/protocol-periphery-v1/compare/725e774409b9137dabcc0776ea4e638640a03db8..61049137d2adb38b0949fbe7c258199d54d6df2e

File: contracts/story-nft/StoryBadge.sol [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    /// @param orgTokenId_ The associated organization NFT token ID.
+    /// @param orgIpId_ The ID of the associated organization IP.
+    /// @param initData The data to initialize the collection.
+    function initialize(
+        address owner_,
+        address orgNft_,
+        uint256 orgTokenId_,
+        address orgIpId_,
+        bytes calldata initData
+    ) public override initializer {
```
[the remark this answers is not in the document]
Comment (sebsadface) [reply contracts/story-nft/StoryBadge.sol]: Currently, the `initialize` function decodes `initData` based on the custom-defined `InitParam` struct in `IStoryBadgeNFT`. If we move this implementation to `BaseStoryNFT`, then the inheriting contracts has to adopt the `InitParam` struct too.

[the remark this answers is not in the document]
Comment (kingster-will) [reply contracts/story-nft/StoryBadge.sol]: I see. we might consider to refactor the parameters structure of `initialize` function. but we can address this in the next change.

File: script/deployment/StoryNFT.s.sol [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+        _readStoryProtocolCoreAddresses();
+        (address defaultLicenseTemplate, uint256 defaultLicenseTermsId) =
+            ILicenseRegistry(licenseRegistryAddr).getDefaultLicenseTerms();
+        address storyNftFactorySigner = vm.envAddress("STORY_NFT_FACTORY_SIGNER");
+        address rootOrgNftRecipient = vm.envAddress("ROOT_ORG_NFT_RECIPIENT");
+        address rootStoryNftOwner = vm.envAddress("ROOT_STORY_NFT_OWNER");
+        address rootStoryNftSigner = vm.envAddress("ROOT_STORY_NFT_SIGNER");
+        string memory rootOrgName = "Test Root Org";
+        string memory rootOrgTokenURI = string(abi.encodePacked(
```
Comment (kingster-will) [inline script/deployment/StoryNFT.s.sol:28]: What's the purpose of the script? we should not deploy such Test Org in product env, sounds right?

Comment (sebsadface) [reply script/deployment/StoryNFT.s.sol:28]: Yeah production shouldn't use these values. I used this script to deploy test contracts.

Comment (kingster-will) [reply script/deployment/StoryNFT.s.sol:28]: In that case, the file name should be renamed to `StoryNFTTest.s.sol` for better readability.

I didn't find a script to only for deploying StoryNFT related contracts, should we add such script?

File: script/utils/DeployHelper.sol [hunk 9]
```diff
@@ -160,7 +172,112 @@ contract DeployHelper is
         }
     }
 
-    function _deployPeripheryContracts() private {
+    function _deployAndConfigStoryNftContracts(
```
Comment (sebsadface) [inline script/utils/DeployHelper.sol:175]: @kingster-will The deploy logic for Story NFT related contracts are here in `DeployHelper`. `StoryNFT.s.sol` just calls this function.

File: script/deployment/StoryNFT.s.sol (hunk 8 above)
Comment (sebsadface) [reply script/deployment/StoryNFT.s.sol:28]: I can take out the root org deployment logic and handle that with a local script. Then, we’ll just use `StoryNFTTest.s.sol` to deploy `StoryNFTFactory` and `OrgNFT`.

Comment (kingster-will) [reply script/deployment/StoryNFT.s.sol:28]: I would suggest to remove all test logic and keep StoryNFT.s.sol as deployment script to deploy official StoryNFT Factory and template contracts.

Comment (sebsadface) [reply script/deployment/StoryNFT.s.sol:28]: sounds good! will make the changes :)

Comment (sebsadface) [reply script/deployment/StoryNFT.s.sol]: Resolved in [c644e86](https://github.com/storyprotocol/protocol-periphery-v1/pull/87/commits/c644e8656d81fa0b132d160d6ea5aacad3dcaf4a) :)

File: script/utils/DeployHelper.sol (hunk 9 above)
Comment (sebsadface) [reply script/utils/DeployHelper.sol:175]: Resolved in [c644e86](https://github.com/storyprotocol/protocol-periphery-v1/pull/87/commits/c644e8656d81fa0b132d160d6ea5aacad3dcaf4a) :)

**Review by kingster-will — APPROVED** [review]
LGTM.

**Pull request merged**